### ISSUE TITLE
unable to decrypt TLS traffic due to ephemeral dh

### ISSUE THREAD

**Issue by osysltd** [bettercap/bettercap#559]
### Environment
* Bettercap version latest just compliled
* OS version Ubuntu SMP Tue Apr 16 08:09:09 UTC 2019 x86_64
* go version go1.10.4 linux/amd64

```set net.sniff.local true
set net.sniff.filter tcp port 443 or tcp port 80
set net.sniff.output capture.cap
```
First of all, thanks for great job, more than well done! Heads up!!

I was playing with captures and can see that the TLS communication with bettercap https.proxy is based on TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 an ephemeral Diffie-Hellman cipher suite, a hard way to decrypt. Could we just downgrade to RSA so that the capture could be easily decrypted using bcap private key? Or I'm missing smth?

**Comment by yungtravla** [comment]
If you have already installed bettercap's root CA on the victim's machine or the victim has temporarily allowed an unsafe connection, there is no need to downgrade TLS since you have the key.

**Comment by osysltd** [comment]
Sorry but bcap uses ephemeral dh and I cannot dectypt the captured communication using the bcap key

**Comment by yungtravla** [comment]
I'm not sure if weakening the encryption is the best way to deal with this issue. If you have no issues proxying TLS and HTTPS traffic then bettercap is already taking care of all the key exchanges. 

> DHE (DH in ephemeral mode) is used in conjunction with a signing certificate, which can be either RSA or DSA. The server sends its cert which client validates, and also its DH group and ephemeral publickey signed under its cert; client generates its ephemeral key in the same group and sends that, plus its cert and signature if client authentication is used.

Correct me if I'm wrong but I think what you are asking for is where to find the keys generated for the ephemeral key exchanges?

The easiest solution I can think of now is to save the HTTPS packets using a `https.proxy.script`, because I can imagine having to save key exchanges for each client and each server is going to get bulky.

**Comment by osysltd** [comment]
Yes, you're right. Any ideas on implementation part?

**Comment by evilsocket** [comment]
@yungtravla since we're proxying stuff, weaking encryption is pretty much our purpose. @osysltd thanks for the report, this is indeed a bug.

**Comment by evilsocket** [comment]
@osysltd i'm not sure i understand: why can't you just use the `https.proxy` with your own CA? the `net.sniff` module operates at the packet level, rebuilding all TLS streams and replacing the key exchance is not doable in a protocol agnostic (or efficient) way.

Nowhere in the documentation we wrote that you can decrypt `net.sniff` captured TLS traffic with bettercap's CA (which is only used by the proxy modules at the application layer)

**Comment by osysltd** [comment]
@evilsocket thanks for catching up the discussion. I don't see much difference between bcap ca or any other one. Nevertheless if you will take a look at any net.sniff capture for the https.proxy you will see ephemeral dh. What Im trying to reach is to have pcap for https decrypted. If this is not possible that the only way to get credentials is by using proxy script which must cover tons of credential input scenarios and will not work for ftps for example.

**Comment by evilsocket** [comment]
no, the proxy script only needs to dump the whole request and you then need to grep on its output ... instead of searching in a pcap ... i don't see a lot of difference to be honest, especially if you're only interested in http traffic, but i'll leave this open as a feature request.

If i get it right, you want us to change how bettercap's certificates are generated on the fly, disabling ephimeral keys, so you can `net.sniff` at the same time and decrypt with that CA the pcap things, correct?

**Comment by osysltd** [comment]
Yep, the main idea here is to stick to cipher suites using the RSA key exchange for the proxy so that the private RSA bcap key could be used to decrypt the encrypted pre-master secret and as a result the whole captured pcap with ease.

Ref:
https://jimshaver.net/2015/02/11/decrypting-tls-browser-traffic-with-wireshark-the-easy-way/
https://security.stackexchange.com/questions/35639/decrypting-tls-in-wireshark-when-using-dhe-rsa-ciphersuites

**Comment by Rubbonn** [comment]
Just to point out that there is an environment variable called SSLKEYLOGFILE that points to a file where browsers and applications writes the keys generated from this process.
It would be nice if the https.proxy module export the keys there.
This is already supported by wireshark and will be easier to decrypt https traffic than a dumper script.

Sources:
[https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/Key_Log_Format](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/Key_Log_Format)
[https://wiki.wireshark.org/TLS](https://wiki.wireshark.org/TLS)
